File: julia/compute_heavy.jl
function matrix_determinant(size::UInt64)::UInt64
    n = size + 2
    matrix = Matrix{UInt64}(undef, n, n)

    for i in 1:n
        for j in 1:n
            matrix[i, j] = ((i-1) * (j-1) + 1) % 1000
        end
    end

    det = UInt64(1)
    for i in 1:n
        for j in 1:n
            det = (det * matrix[i, j]) % 1000000
        end
    end
    return det
end

function factorial(n::UInt64)::UInt64
    if n == 0 || n == 1
        return UInt64(1)
    end
    result = UInt64(1)
    for i in 2:n
        result = result * i
    end
    return result
end

function hash_compute(seed::UInt64)::UInt64
    hash = seed
    for _ in 1:100
        hash = hash * 1103515245 + 12345
        hash = (hash >> 16) ⊻ hash
    end
    return hash
end

function main()
    start_time = time_ns()

    result = UInt64(0)
    iterations = 1_000_000

    for i in 0:(iterations-1)
        result += matrix_determinant(UInt64(i % 15))
        result += factorial(UInt64(i % 18))
        result += hash_compute(UInt64(i))
    end

    duration = time_ns() - start_time
    println(duration)
    println("Result: ", result)
    exit(0)
end

main()

File: nodejs/compute_heavy.js
const MAX_U64 = 0xffffffffffffffffn;

function uint64(n) {
  return BigInt(n) & MAX_U64;
}

function matrixDeterminant(size) {
  const n = size + 2;
  const matrix = Array(n)
    .fill(null)
    .map(() => Array(n).fill(0));

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      matrix[i][j] = (i * j + 1) % 1000;
    }
  }

  let det = 1n;
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      det = (det * BigInt(matrix[i][j])) % 1000000n;
    }
  }
  return det;
}

function factorial(n) {
  if (n === 0 || n === 1) {
    return 1n;
  }
  let result = 1n;
  for (let i = 2; i <= n; i++) {
    result = uint64(result * BigInt(i));
  }
  return result;
}

function hashCompute(seed) {
  let hash = BigInt(seed);
  for (let i = 0; i < 100; i++) {
    hash = uint64(hash * 1103515245n + 12345n);
    hash = uint64((hash >> 16n) ^ hash);
  }
  return hash;
}

function main() {
  const start = process.hrtime.bigint();

  let result = 0n;
  const iterations = 1_000_000;

  for (let i = 0; i < iterations; i++) {
    result = uint64(result + matrixDeterminant(i % 15));
    result = uint64(result + factorial(i % 18));
    result = uint64(result + hashCompute(i));
  }

  const duration = process.hrtime.bigint() - start;
  console.log(duration.toString());
  console.log("Result:", result.toString());
  process.exit(0);
}

main();
